_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,233 @@
+---
+layout: post
+title:  "Migrate your Strimzi-operated cluster from ZooKeeper to KRaft"
+date: 2024-03-20
+author: paolo_patierno
+---
+
+In the previous [blog post](TBD) we explored how ZooKeeper is used to store Kafka clusters metadata and what are the current limitations.
+We introduced the new KRaft protocol which allows to overcome such limitations and store metadata in Kafka itself, taking into account that the ZooKeeper support will be removed soon.
+For this reason, there is the need to migrate existing clusters from using ZooKeeper to KRaft for storing metadata.
+We also covered the overall procedure to do that.
+Such procedure is not that simple because of the several configuration changes to be made on the cluster nodes, as well as the multiple times the nodes have to be rolled.
+But if your cluster is running on Kubernetes and it is Strimzi-operated, we got you covered!
+In this blog post, we are going to see how the Strimzi cluster operator provides a semi-automated process to run the migration easily, forgetting all the manual steps you should be doing to reach the same goal.
+
+<!--more-->
+
+### Strimzi migration support
+
+The migration procedure is not that easy and it relies on updating brokers and controllers configurations multiple times together with rolling the involved nodes across several phases.
+There is a lot of manual intervention even related to verify the migration status by looking at the KRaft controller leader log or checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric.
+But if you are using the Strimzi operator to run your current ZooKeeper-based cluster on Kubernetes, the migration process is semi-automated.
+What you have to do is just updating the `strimzi.io/kraft` annotation on the `Kafka` cluster resource in order to allow the internal migration Finite State Machine (FSM) to move across the different phases.
+The operator is going to apply the needed configuration to the nodes and roll them.
+It is also in charge of checking the migration status looking at the metrics for you to make sure it was done successfully.
+
+Before latest Strimzi 0.40.0 release, the `strimzi.io/kraft` annotation on the `Kafka` custom resource was used to define a ZooKeeper-based cluster, by using the `disabled` value, or a KRaft-based cluster, by using the `enabled` value.
```

Comment:
Screw the past ... I would just say that the migration is supported from Strimzi 0.40.0 and not focus on how things worked before.

Reply:
Tbh I think it's useful making the point that the `kraft` annotation is not new and the current values are used to define ZooKeeper or KRaft based clusters. Then we used the same annotation for the migration with two additional values.